### PR TITLE
Bringback some MPI.h items removed by 89da9651b

### PR DESCRIPTION
**Pull request by gpaulsen**
In commit 89da9651b, we finally removed some MPI items that were
removed in MPI 3.0 / 3.1 (Unless the user configures with an
mpi1-compat flag).  Unfortunately some of the items removed in
that commit, were NOT removed in the standard, and were left as
deprecated.

This commit brings those items back, as deprecated with/without
the --mpi1-compat flag.

Signed-off-by: Geoffrey Paulsen <gpaulsen@us.ibm.com>

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/1b59d2583b7bd4427ec4254a72828edd

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/a7582345f4a5782039037cacc9d4de21

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/804a16255aca0fced2cdb7144ea6de2f

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/206058d885e6e4095c190040850f81d3

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/10b0d4ed20050fb0b268b948ed8f5d31

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ea9ae5d8f0861433ac8a873fb2c6b915

File: ompi/include/mpi.h.in [hunk 1]
```diff
@@ -406,28 +404,15 @@ typedef int (MPI_Grequest_query_function)(void *, MPI_Status *);
 typedef int (MPI_Grequest_free_function)(void *);
 typedef int (MPI_Grequest_cancel_function)(void *, int);
 
-#if !OMPI_OMIT_MPI1_COMPAT_DECLS
 /*
- * Removed typedefs. These will be deleted in a future Open MPI release.
+ * Deprecated typedefs. These will be deleted in a future MPI standard.
```
Comment (jsquyres) [inline ompi/include/mpi.h.in]: These *may* be deleted in a future version of the MPI standard.

"Deprecated", as defined by MPI, does *not* mean "will be deleted in a future version of the MPI standard."

File: ompi/include/mpi.h.in [hunk 2]
```diff
@@ -806,10 +791,15 @@ enum {
    protection for this case. */
 
 #if !defined(OMPI_COMPILING_FORTRAN_WRAPPERS)
+/*
+ * Deprecated Defines:
```
Comment (jsquyres) [inline ompi/include/mpi.h.in]: Can you put the MPI spec version when these were deprecated?

Also, the MPI spec doesn't say whether these are macros or not.  So they're typically not "Deprecated defines".

File: ompi/include/mpi.h.in [hunk 3]
```diff
@@ -806,10 +791,15 @@ enum {
...
+ */
 #define MPI_NULL_DELETE_FN OMPI_C_MPI_NULL_DELETE_FN
 #define MPI_NULL_COPY_FN OMPI_C_MPI_NULL_COPY_FN
 #define MPI_DUP_FN OMPI_C_MPI_DUP_FN
 
+/***/
```
Comment (jsquyres) [inline ompi/include/mpi.h.in]: What is this for?

File: ompi/include/mpi.h.in [hunk 4]
```diff
@@ -2624,89 +2595,135 @@ OMPI_DECLSPEC  int MPI_T_pvar_readreset(MPI_T_pvar_session session, MPI_T_pvar_h
...
-OMPI_DECLSPEC  int PMPI_Type_lb(MPI_Datatype type, MPI_Aint *lb)
-                                __mpi_interface_removed__("MPI_Type_lb has been removed, use MPI_Type_get_extent in MPI-2.0");
+            __mpi_interface_removed__("MPI_Type_struct is removed in MPI-3.0; use MPI_Type_create_struct instead.");
 OMPI_DECLSPEC  int PMPI_Type_struct(int count, int array_of_blocklengths[],
                                     MPI_Aint array_of_displacements[],
                                     MPI_Datatype array_of_types[],
                                     MPI_Datatype *newtype)
-                                    __mpi_interface_removed__("MPI_Type_struct is superseded by MPI_Type_create_struct in MPI-2.0");
+            __mpi_interface_removed__("PMPI_Type_struct is removed in MPI-3.0; use PMPI_Type_create_struct instead.");
+/***/
```
Comment (jsquyres) [inline ompi/include/mpi.h.in]: What's this for?

File: ompi/include/mpi.h.in [hunk 5]
```diff
@@ -2624,89 +2595,135 @@ OMPI_DECLSPEC  int MPI_T_pvar_readreset(MPI_T_pvar_session session, MPI_T_pvar_h
...
+OMPI_DECLSPEC  int MPI_Type_ub(MPI_Datatype mtype, MPI_Aint *ub)
+            __mpi_interface_removed__("MPI_Type_ub has been removed in MPI-3.0; use MPI_Type_get_extent instead.");
 OMPI_DECLSPEC  int PMPI_Type_ub(MPI_Datatype mtype, MPI_Aint *ub)
-                                __mpi_interface_removed__("MPI_Type_ub has been removed, use MPI_Type_get_extent in MPI-2.0");
+            __mpi_interface_removed__("PMPI_Type_ub has been removed in MPI-3.0; use PMPI_Type_get_extent instead.");
+OMPI_DECLSPEC  int MPI_Type_lb(MPI_Datatype type, MPI_Aint *lb)
+            __mpi_interface_removed__("MPI_Type_lb has been removed in MPI-3.0; use MPI_Type_get_extent instead.");
+OMPI_DECLSPEC  int PMPI_Type_lb(MPI_Datatype type, MPI_Aint *lb)
+            __mpi_interface_removed__("PMPI_Type_lb has been removed in MPI-3.0; use PMPI_Type_get_extent instead.");
+/***/
```
Comment (jsquyres) [inline ompi/include/mpi.h.in]: What's this for?

File: ompi/include/mpi.h.in (hunk 3 above)
Comment (gpaulsen) [reply ompi/include/mpi.h.in]: There are lines in the MPI document separating these sections of them, so I added this.  I can remove.

File: ompi/include/mpi.h.in (hunk 4 above)
Comment (gpaulsen) [reply ompi/include/mpi.h.in]: again, maps to a line in the MPI 3.1 standard table 2.1 page 18

File: ompi/include/mpi.h.in (hunk 5 above)
Comment (gpaulsen) [reply ompi/include/mpi.h.in]: same.

**Comment by gpaulsen** [comment]
@jsquyres  thanks for the review.  I'll continue to itereate.  Please respond to my replies to your questions, and I'll get those in too.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/0a533d0b0803c0b28450d802234bfeca

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/00376c2a9be65e0b785ad28ee131b4c3

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/24ac67f34d9cf2dce4d7b1273490d00e

**Comment by gpaulsen** [comment]
I accidentally deleted my fork's branch associated with this PR, which closed this.
The new / better version of this PR is here: https://github.com/open-mpi/ompi/pull/6013

**Pull request closed without merging**